pkg/sdn/plugin/dns.go

```diff
@@ -0,0 +1,186 @@
+package plugin
+
+import (
+	"fmt"
+	"net"
+	"os/exec"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+
+	kerrors "k8s.io/kubernetes/pkg/util/errors"
+)
+
+const (
+	DIG = "dig"
+
+	defaultTTL = 30 * time.Minute
+)
+
+type dnsValue struct {
+	// All IPv4 addresses for a given domain name
+	ips []net.IP
+	// Time-to-live value from non-authoritative/cached name server for the domain
+	ttl time.Duration
+	// Holds (last dns lookup time + ttl), tells when to refresh IPs next time
+	nextQueryTime time.Time
+}
+
+type DNS struct {
+	// Protects dnsMap operations
+	lock sync.Mutex
+	// Holds dns name and its corresponding information
+	dnsMap map[string]dnsValue
+}
+
+func CheckDNSResolver() error {
+	if _, err := exec.LookPath(DIG); err != nil {
+		return fmt.Errorf("%s is not installed", DIG)
+	}
+	return nil
+}
+
+func (d *DNS) Size() int {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	return len(d.dnsMap)
+}
+
+func (d *DNS) Get(dns string) dnsValue {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	data := dnsValue{}
+	if res, ok := d.dnsMap[dns]; ok {
+		data.ips = make([]net.IP, len(res.ips))
+		copy(data.ips, res.ips)
+		data.ttl = res.ttl
+		data.nextQueryTime = res.nextQueryTime
+	}
+	return data
+}
+
+func (d *DNS) Add(dns string) error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	if len(d.dnsMap) == 0 {
+		d.dnsMap = map[string]dnsValue{}
+	}
+	d.dnsMap[dns] = dnsValue{}
+	err, _ := d.update(dns)
+	return err
+}
+
+func (d *DNS) Update() (error, bool) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	errList := []error{}
+	changed := false
+	for dns := range d.dnsMap {
+		err, updated := d.update(dns)
+		if err != nil {
+			errList = append(errList, err)
+			continue
+		}
+		if updated {
+			changed = true
+		}
+	}
+	return kerrors.NewAggregate(errList), changed
+}
+
+func (d *DNS) update(dns string) (error, bool) {
+	out, err := exec.Command("sh", []string{"-c", digShellCmdGetIPAndTTL(dns)}...).Output()
```

Comment:
Could we get some unit tests for this function?  Especially since it's parsing command output.